### PR TITLE
Feature: Plot layout images

### PR DESCRIPTION
**Pull request by mdtusz**
Continuing from before - fulfilling https://github.com/plotly/plotly.js/issues/496:

This PR adds the ability to add arbitrary images loaded from a url to a plot. Plot layout images follow a similar interface to shapes, but use a coordinate position with `sizex` and `sizey` rather than `x0`, `x1`, `y0`, `y1`. Additionally, there are three sizing modes - `contain` (the default), `fill`, and `stretch`.

The intended use of this feature is mainly for logos and non-data-specific images - for image data that is tied directly to data (e.g. satellite imagery) we should/will add a new trace type to accommodate for this and enable features such as toggling from the legend and ordering.

_That said_, images _can_ be added to the plot area and be forced to keep their binding to the underlying cartesian plane by setting the `xref` and `yref` to their respective axes and setting the `sizing` property to `stretch`.

Since [the initial PR](https://github.com/plotly/plotly.js/pull/520), a few things have been changed:

* Image loading adds a promise to `gd._promises`
* A key function has been added for relayout updates to images
* `width`/`height` have been changed to `sizex`/`sizey`
* More tests
* A wicked-awesome baseline image 🎉

### CODE REVIEW

File: src/plots/layout_attributes.js [hunk 1]
```diff
@@ -182,6 +182,7 @@ module.exports = {
         'legend': 'Legend',
         'annotations': 'Annotations',
         'shapes': 'Shapes',
+        'images': 'Images',
```
Comment (etpinard) [inline src/plots/layout_attributes.js]: @mdtusz can you also add `images` to this test [list](https://github.com/plotly/plotly.js/blob/19030614649428a0341d28402178bdd32fc75357/test/jasmine/tests/plotschema_test.js#L118) ?

File: src/components/images/attributes.js [hunk 2]
```diff
@@ -37,24 +37,24 @@ module.exports = {
         ].join(' ')
     },
 
-    width: {
+    sizey: {
```
Comment (etpinard) [inline src/components/images/attributes.js]: `sizex` maybe ?

File: src/plot_api/plot_api.js [hunk 3]
```diff
@@ -2299,6 +2299,12 @@ Plotly.relayout = function relayout(gd, astr, val) {
             // as it is we get separate calls for x and y (or ax and ay) on move
             objModule.draw(gd, objNum, p.parts.slice(2).join('.'), aobj[ai]);
             delete aobj[ai];
+        } else if(p.parts[0] === 'images') {
+            var update = Lib.toObject(astr, vi);
```
Comment (etpinard) [inline src/plot_api/plot_api.js]: very nice.

File: src/components/images/draw.js [hunk 4]
```diff
@@ -121,16 +145,22 @@ module.exports = function draw(gd) {
     }
 
 
+    // Required for updating images
+    function keyFunction(d) {
```
Comment (etpinard) [inline src/components/images/draw.js]: well played.

File: test/jasmine/tests/layout_images_test.js [hunk 5]
```diff
@@ -227,4 +229,47 @@ describe('Layout images', function() {
...
+        });
+
+        it('should remove the image tag if an invalid source', function(done) {
+
+            var selection = Plotly.d3.select('image');
+            expect(selection.size()).toBe(1);
+
+            Plotly.relayout(gd, 'images[0].source', 'invalidUrl').then(function() {
+                var newSelection = Plotly.d3.select('image');
+                expect(newSelection.size()).toBe(0);
```
Comment (etpinard) [inline test/jasmine/tests/layout_images_test.js]: great test

File: src/components/images/draw.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+
+            thisImage.on('error', errorHandler);
+            thisImage.on('load', resolve);
+
+            img.src = d.source;
+
+            function errorHandler() {
+                thisImage.remove();
+                resolve();
```
Comment (etpinard) [inline src/components/images/draw.js]: Yeah, I think you got this right. Throwing an error on incorrect `source` input would be a little bit too much I think.

**Comment by etpinard** [comment]
@mdtusz Looking great. Here are my TODOs:

- [ ] https://github.com/plotly/plotly.js/pull/525#discussion_r63037797
- [ ] I think you got `sizex` and `sizey` mixed up in the [attributes](https://github.com/plotly/plotly.js/pull/525/files#diff-aff79bb973d6b604c1efc0b7fa923fddR40)
- [ ] make sure that we are not vulnerable to XSS [here](https://github.com/plotly/plotly.js/pull/525/files#diff-43b5cc6f9312074b8335172520f2cb6eR83)
- [ ] I think the image `draw` module is missing some logic to handle `xanchor` and `yanchor` to be on-par with annotations ([ref](https://github.com/plotly/plotly.js/blob/19030614649428a0341d28402178bdd32fc75357/src/components/annotations/index.js#L405-L418)). 
- [ ] add a few test cases for [`Lib.toObject`](https://github.com/plotly/plotly.js/pull/525/files#diff-e64ef427412e9bab4929b8b93ca12d96R506) in `lib_test.js`
- [ ] Looks like images are _always_ [drawn](https://github.com/plotly/plotly.js/pull/525/files#diff-2941ab69a12080c0633ff4ac8ea3aa83R2821) above shapes. Not sure what's best. @mdtusz what's your rationale on this topic?
- [ ] we should add one `data:image` image to the test json [mock](https://github.com/plotly/plotly.js/pull/525/commits/5b9077c4ca500a79968a620da7ef053d6af1f50c)

**Comment by mdtusz** [comment]
I'm not too sure how best to deal with shapes/images layering. I'd prefer if we kept it absolute so there's less convoluted logic on the layering though (and also, it would get pretty messy trying to define the order they are drawn). Images are on top right now purely because of the order their `<g>` containers are added though, so I'm happy to swap them - might make more sense so people can draw things atop the images for whatever reason.

**Comment by etpinard** [comment]
> I'd prefer if we kept it absolute so there's less convoluted logic on the layering though

good call.

> might make more sense so people can draw things atop the images for whatever reason.

+1. Let's go with that.

File: src/lib/index.js [hunk 7]
```diff
@@ -495,15 +495,15 @@ lib.isIE = function() {
...
+ * lib.objectFromPath('nested.test[2].path', 'value');
  * // returns { nested: { test: [null, null, { path: 'value' }]}
  *
  * @param   {string}    path to nested value
  * @param   {*}         any value to be set
  *
  * @return {Object} the constructed object with a full nested path
  */
-lib.toObject = function(path, value) {
+lib.objectFromPath = function(path, value) {
```
Comment (etpinard) [inline src/lib/index.js]: an even more verbose _plotly-esque_ name would be `Lib.objectFromAttributeString` or `Lib.objectFromAstr`.

@mdtusz I'll leave it up to you to decide what's best.

File: test/image/mocks/layout_image.json [hunk 8]
```diff
@@ -39,19 +38,32 @@
...
...ze46iGYz50spozKgyOuZagX3Ztybw96V99EF4yhxNoBUNOQmxsQraVG+Qsc2r60ldNppfb/nwiycevJIGxbst7r5tMExavGRnFhoajXaHeHN6PDEHzURjuz35auAAUgOP4lh+AFuGicVNVXEFV/wIdbgCa1/pZGoQDY00xVaRFuBkY+0y5rPIQZCDDARzCFzfapkW4s5j+LF+i9mhuAMOBgZ0Ymp2lPTiHnx2GfuB+1d8xBCwc6Huwk4TjkjUDB/pEq/ArkImdTOh42VesUt23wIM7tf3+oTrrvp0qbh/ywkFDcJ5/3yh3wlN8/hvgcCebPM0nyAENl1aj9lP9xGzgwZ3bzUIR8YvGd74I8HR1RYSHhy+pG0DWMqHsz44uhTdM7Qv0N6JISEhISEhISEhISEhISEieg/wHXMU13xThNeUAAABhdEVYdGNvbW1lbnQARmlsZSBzb3VyY2U6IGh0dHA6Ly9jb21tb25zLndpa2ltZWRpYS5vcmcvd2lraS9GaWxlOlBsb3RseV9sb2dvX2Zvcl9kaWdpdGFsX2ZpbmFsXyg2KS5wbmetqmbcAAAAJXRFWHRkYXRlOmNyZWF0ZQAyMDE0LTA5LTIxVDAzOjM1OjExKzAwOjAwKhOa5QAAACV0RVh0ZGF0ZTptb2RpZnkAMjAxNC0wOS0yMVQwMzozNToxMSswMDowMFtOIlkAAABGdEVYdHNvZnR3YXJlAEltYWdlTWFnaWNrIDYuNi45LTcgMjAxNC0wMy0wNiBRMTYgaHR0cDovL3d3dy5pbWFnZW1hZ2ljay5vcmeB07PDAAAAGHRFWHRUaHVtYjo6RG9jdW1lbnQ6OlBhZ2VzADGn/7svAAAAGXRFWHRUaHVtYjo6SW1hZ2U6OmhlaWdodAAxNzA0VsMjbQAAABh0RVh0VGh1bWI6OkltYWdlOjpXaWR0aAAxNTMxndUDgQAAABl0RVh0VGh1bWI6Ok1pbWV0eXBlAGltYWdlL3BuZz+yVk4AAAAXdEVYdFRodW1iOjpNVGltZQAxNDExMjcwNTEx0Rvi6AAAABN0RVh0VGh1bWI6OlNpemUAMjkuOUtCQvmCzmYAAAAzdEVYdFRodW1iOjpVUkkAZmlsZTovLy90bXAvbG9jYWxjb3B5X2UwNDYwYmM4MDVlYS0xLnBuZ1YfjUoAAAAASUVORK5CYII=",
```
Comment (etpinard) [inline test/image/mocks/layout_image.json]: dope

**Comment by etpinard** [comment]
:dancer:

**Pull request merged**